### ISSUE TITLE
Can't update 'checked' attribute

### ISSUE THREAD

**Issue by jdelafon** [facebook/react#6321]
Consider this input:

```
<input
    type='radio'
    value='2'
    checked={this.state.value === '1'}
    onChange={this.onChange}
/>

onChange: function (e) {
    this.setState({value: e.target.value});
}
```

If the first time the component is rendered, `this.state.value === '1'`, then this input will remain `checked` forever in the DOM, even if it appears unchecked visually. For instance this can happen with

```
getInitialState = function() {
    return {value: '1'};
}
```

I am using React 0.14.7.

**Comment by zpao** [comment]
I can't reproduce this (or maybe am misunderstanding what you're saying). https://jsfiddle.net/rkc64ybp/ is showing that the node is unchecked in the DOM. You might run into issues if you use `e.preventDefault()` but I don't see that in your code here.

**Comment by Aweary** [comment]
@zpao I believe he means that, even after unchecking it, the `<input>` element still has the `checked` attribute. That seems to be the case on my end, at least. Chrome Version 49.0.2623.87 (64-bit) FWIW.

**Comment by muraveill** [comment]
Yes, that is what I meant, sorry if it wasn't clear. I tried Firefox and Chrome. 
It is not really a problem for user experience, but it is for functional testing (although a workaround is to set `className={value === 1 ? 'checked' : ''}` and use the class as a selector).

**Comment by zpao** [comment]
Ah, I see. The attribute is set but the property isn't. That will be solved in v15 where we will remove the attribute instead of just setting the property to false. See https://jsfiddle.net/rkc64ybp/1/ (same code, just with `hasAttribute` calls and upgraded to 15, I saw the issue with 0.14)

**Comment by Aweary** [comment]
@zpao should React add the `checked` attribute to the `<input>` element when it's checked in `v0.15`? With that JSFiddle it just looks like it doesn't add the `checked` attribute at all.

**Comment by zpao** [comment]
That's a good question… @syranide - was there any discussion about that?

**Comment by syranide** [comment]
@zpao There was/is a separate discussion about inputs and value somewhere else (cc @spicyj). I'm not sure where it ended up but it was about going away from the native behavior seen here to actively update the attribute in the DOM. However, it should be noted that the behavior seen here **is NOT incorrect behavior**, you get this behavior with native checkboxes as well; the `checked` property is updated, the attribute is not as it reflects the initial state.

**Comment by muraveill** [comment]
Ah, I did not know that it was the native behavior... Then it is not a React issue at all. That sucks though, it means for instance that one should never use the `input:checked` CSS selector anyway to style the currently selected element, and a className='checked' is the only way Selenium can find it.

**Comment by syranide** [comment]
@muraveill I would assume `elem:checked` works and that's specifically why it was added, but not `elem[checked]` (tests for attribute).

**Comment by spicyj** [comment]
Once again, I'll repeat that it does not make sense to talk about the native behavior here. That is, HTML does not have a concept of going from

```
<input type="checkbox">
```

to

```
<input type="checkbox" checked>
```

We created this concept. In HTML, you can do a property assignment (`.value =`), attribute assignment (`.setAttribute`), or you can blow away the entire thing and replace the whole node with `.innerHTML`. You can also have the user click it.

We can decide for any of these to be the behavior that React matches. None of them is the objectively correct "native behavior" so it's misleading to talk as if that were the case.

It seems like setting the attribute is least confusing for people in most cases (so that both `[checked]` and `:checked` would work), but it's still not clear whether that's best, especially with forms.

**Comment by syranide** [comment]
> Once again, I'll repeat that it does not make sense to talk about the native behavior here. That is, HTML does not have a concept of going from

> We can decide for any of these to be the behavior that React matches. None of them is the objectively correct "native behavior" so it's misleading to talk as if that were the case.

I still don't buy it, the specification explicitly states that the attribute reflects the initial value and the property reflects the current value. So if you check it without any additional scripting applied, this is the behavior you get, you're also meant to generally update it via `.checked` (then it's consistent with user interaction). What's being asked here is not something you can do with plain native checkboxes, so if React does something that enables it, how is that not deviating? But yes, React doesn't really fit the HTML model perfectly so it's fine, but I think there's quite clearly a native behavior.

Anyway, I don't want to derail this.

**Comment by spicyj** [comment]
Why are you "meant to" update it with the property? `.setAttribute()` is an equal here.

**Comment by syranide** [comment]
If you want to simulate user input/filter/normalize/etc then you use the property so you don't change the initial value. If you want to replace  (i.e. changing data source) the value then you use the attribute (similar to setting the key in React), you're "resetting" the checkbox in a sense. Again I agree this differentiation doesn't really work well in React (and there's no good reason to).

I don't really mind, it doesn't really have any major implications, but it is clearly different from plain IMHO.

**Comment by jimfb** [comment]
Conceptually, upon every change, we throw away the entire page and rerender from scratch.  Ideally, the output of an initial render should be identical to the output of an initial render followed by an update render, assuming the final state is the same.

**Comment by syranide** [comment]
@jimfb Yes, but it's a partial truth, inputs also have focus, selection, undo history, etc and in the context of this discussion, they also have an initial value that is reflected in the attribute. We currently control the lifetime of internal and external state via `key` in React and it makes sense, even for seemingly controlled components. If you don't agree with HTMLDOM reflecting the initial value in the attribute that is fine, it's a weird HTML quirk, but if interacting with a plain checkbox (attribute not updated) and a React checkbox (attribute updated) produces different results then React is doing something differently, period. Again, IMHO that's fine, but it is clearly deviating from normal HTMLDOM behavior.

**Comment by muraveill** [comment]
Without any consideration to a "native" behavior, what I understand is that the "checked" attribute of an input in a reactive interface, indicating the initial state, is a totally useless thing. Is there any way then that React could make it so that setting `myinput.checked = "checked"` would also check the item visually ? 

Example: let's say that my `onChange` callback triggers a long action like a computation on the server. I want my input to update now to appear checked, not only when the computation is finished - otherwise the user thinks he has not clicked right and keeps clicking. A simple way would be to do `myinput.checked = "checked"` as suggested above. The current way is write 10 ugly and stupid lines of code like theses ones so that a state updates first, and then we get the props at the end of the action:

        getInitialState: function () {
            // "anti-pattern", according to the docs
            return {value: this.props.value};
        },
        componentWillReceiveProps: function(nextProps) {   
            // because we use the state in the <input>, it needs to mirror the props. I call it dirty.
            this.setState({value: nextProps.value});
        },
        onChange: function (e) {
            this.setState({value: e.target.value});   // to update it right now
            Actions.updateOneValue(e.target.value);   // long action
        },

        render: function() {
            ...
            return <input checked={ this.state.value === 1} />;
        }

**Comment by rovolution** [comment]
had a breaking change in one of my unit test suites regarding this issue when upgrading from 0.14 -> 0.15...I built a custom checkbox component that wraps an `<input type="checkbox" />`. Personally, I prefer when it sets the `"checked"` attribute on the `<input>` element

**Comment by krzysu** [comment]
I agree with @rovolution. I see similar case in my code, I use radio inputs like this (a few of them):
```
<input
    type="radio"
    name={attributeName}
    checked={config.selected}
    (...)
/>
```
on every change value of `config.selected` is changed. In the browser it works correctly, but in dev mode I'm getting this in the console:

```
Warning: React attempted to reuse markup in a container but the checksum was invalid. This generally means that you are using server rendering and the markup generated on the server was not what the client was expecting. React injected new markup to compensate which works but you have lost many of the benefits of server rendering. Instead, figure out why the markup being generated is different on the client or server:
 (client) <input type="radio" name="color"
 (server) <input type="radio" checked="" name="color"
```
what means the way you render on the server is different than on the client.
after @syranide: "the `checked` property is updated, the attribute is not as it reflects the initial state" <- from what I see the attribute is removed on the client

**Comment by syranide** [comment]
@krzysu `<input type="radio" name="color"` means that it is unchecked and `<input type="radio" checked="" name="color"` means that it is checked. You're rendering different markup on server vs client hence the error/warning.

**Comment by krzysu** [comment]
yes, I know that, but that happens only after upgrading to react v15. so the way new react is rendering inputs is different between client and server, and that's why I'm writing here.
Client removes `checked` attribute (because of new logic) causing checksum to be invalid, causing warning.

**Comment by syranide** [comment]
@krzysu I'm unable to reproduce this myself. Rendering a checked radio input on the server and then resuming on the client works just fine for me, no mismatch error.

**Comment by krzysu** [comment]
ok, I will try to prepare an example which reproduces it

**Comment by rovolution** [comment]
Here are some GIFs with my custom checkbox:

**React 0.14.0 (setting `checked` attribute):**
![with-checked](https://cloud.githubusercontent.com/assets/437318/15020456/1f31ccd0-11f0-11e6-9c42-60e2db4d635b.gif)

**React 0.15.0 (`checked` attribute not set):**
![wihout-checked](https://cloud.githubusercontent.com/assets/437318/15020569/8bc6f9c4-11f0-11e6-90e1-85b58cfdb1cb.gif)

In the code (jsx), I am just setting the `checked` attribute to `true` on an <input/> element of `type="checkbox"`

**Comment by krzysu** [comment]
ok, so my issue is different that the one from @rovolution
 
@syranide as promised, I extracted my code into a reproducible example, which you can find here https://github.com/krzysztofurbas-home24/react15-input-checked-issue.
I've found so far that the problem is caused by some of the dependencies and very specific version of it. That's why I included npm-shrinkwrap.json. If you remove npm-shrinkwrap, and reinstall node_modules, then the reported issue is gone. So I'm not sure if it's worth debugging :wink:

**Comment by syranide** [comment]
@krzysu :+1: Are you using npm3? npm2 has a problem with its hierarchical dependencies such that some plugins may depend on older versions of React and have those be installed side-by-side. They are not compatible and will generate slightly different markup which can cause the problem seen above, which I also find likely considering that the subsequent attributes are not even the same (order) in client/server output warning. This generally shows itself differently in my experience, so who knows, but some external versioning conflict obviously.

**Comment by CoderK** [comment]
@rovolution 
I have the same issue in react@15.4.2

from component.jsx
```
    render() {
        const { name, label, title } = this.props;
        const buttonClass = classnames(
            "se3-toggle-button",
            `toolbar-button-${name}`,
            "toggle-checkbox"
        );
        const checkboxId = `se3-toolbar-${name}-button`;
        const isChecked = this.state.checked;
        const currentTitle = title[isChecked];

        return (
            <div className={ buttonClass }
                title={ currentTitle }
                data-type={ ButtonTypes.TOGGLE }
                data-name={ name }>
                <input type="checkbox"
                    id={ checkboxId }
                    className="se3-toolbar-checkbox"
                    checked={ isChecked }
                    onChange={ this.handleChange }
                    data-role="item" />
                <label htmlFor={ checkboxId }
                    className="se3-toolbar-label">
                    <span className="se3-toolbar-icon"></span>
                    <span className="se3-toolbar-label-text">{ label }</span>
                </label>
            </div>
        );
    }
```

to html
```
<input type="checkbox" id="se3-toolbar-bold-button" class="se3-toolbar-checkbox" data-role="item" value="on">
```

When I click checkbox, It has been changed checked property internally. But the checked attribute 
 of checkbox element's was not displayed in chrome inspector.

![image](https://cloud.githubusercontent.com/assets/1631348/24846672/f0e88ad0-1df6-11e7-94af-01b3866f5979.png)

**Comment by mstijak** [comment]
I'm seeing the same problem in all browsers. Unfortunately, it's difficult to extract a smaller example reproducing the issue. 
![image](https://cloud.githubusercontent.com/assets/433394/25408114/2f01d732-2a0d-11e7-876d-adfa7d13eae3.png)

My workaround is to assign a different key based on the value, but I really hate it.

**Comment by domenikasd** [comment]
please helpe me
 

    Στις 12:20 π.μ. Τετάρτη, 26 Απριλίου 2017, ο/η Marko Stijak <notifications@github.com> έγραψε:
 

 I'm seeing the same problem in all browsers. Unfortunately, it's difficult to extract a smaller example reproducing the issue.
My workaround is to assign a different key based on the value, but I really hate it.—

**Comment by domenikasd** [comment]
εγκατεστησα αυτο το προγραμμα γ να βλεπω κωδικους κ διαφορα αλλα δν ξερω τι να καν απο δω κ περα.εχω κανει προφιλ .βλεπω το πλαισιο  ( κωδικο θεματα αποθετηρια κ ολα τα υπολοιπα αλλα εγω δν ξερω τιποτα που να παω κ πως να δω τους κωδικους κ ολα τα αλλα ειμαι ελληνιδα κ μου ειναι δυσκολο το προγγραμμα ειναι στα αγγλικα μπορεις να με βοηθησεις? 

    Στις 12:20 π.μ. Τετάρτη, 26 Απριλίου 2017, ο/η Marko Stijak <notifications@github.com> έγραψε:
 

 I'm seeing the same problem in all browsers. Unfortunately, it's difficult to extract a smaller example reproducing the issue.
My workaround is to assign a different key based on the value, but I really hate it.—

**Comment by gaearon** [comment]
I’ll close since there’s a few different things discussed here, and it’s hard for me to say what’s the status on either of them.

Please file a new issue with a reproducing example if you still experience problems related to this.
